skbio/io/format/embedding.py

```diff
@@ -0,0 +1,212 @@
+
+    # From https://en.wikipedia.org/wiki/Hierarchical_Data_Format
+    # Note that Wikipedia specifies: "\211HDF\r\n\032\n" which is an ordinal form:
+    # >>> ord('\211')
+    # 137
+    # >>> ord('\x89')
+    # 137
+    # >>> ord('\032')
+    # 26
+    # >>> ord('\x1a')
+    # 26
+    if magic == b"\x89HDF\r\n\x1a\n":
+        with h5py.File(fh, "r") as h5file:
+            if "embedding" in h5file and "id" in h5file and "idptr" in h5file:
+                return True, {}
+
+    return False, {}
+
+
+@embed.reader(None)
+def _embed_to_generator(
+    fh,
+    constructor=ProteinEmbedding,
+    obj_constructor=Protein,
+    kwargs: dict = {},
+):
+    h5grp = h5py.File(fh, "r")
+    embed_fh = h5grp["embedding"]
+    id_fh = h5grp["id"]
+    idptr_fh = h5grp["idptr"]
+    j = 0
+    n = idptr_fh.shape[0]
+    for i in range(n):
+        idptr = idptr_fh[i]
+        id_ = id_fh[j:idptr]
+        emb = embed_fh[j:idptr]
+        string = str(id_.tobytes().decode("ascii")).replace("\x00", "")
+        j = idptr
+        yield constructor(emb, string, **kwargs)
+
+
+def _embed_to_object(
+    fh,
+    constructor=ProteinEmbedding,
+    obj_constructor=Protein,
+    kwargs: dict = {},
+):
+    h5grp = h5py.File(fh, "r")
+    embed_fh = h5grp["embedding"]
+    id_fh = h5grp["id"]
+
+    # assumes that there is only a single object in the file
+    emb = embed_fh[()].squeeze()
+    id_ = id_fh[()]
+    string = str(id_.tobytes().decode("ascii")).replace("\x00", "")
+    return constructor(emb, string, **kwargs)
+
+
+@embed.reader(ProteinEmbedding)
+def _embed_to_protein(
+    fh,
+    kwargs: dict = {},
+):
+    return _embed_to_object(
+        fh,
+        constructor=ProteinEmbedding,
+        obj_constructor=Protein,
+        kwargs=kwargs,
+    )
+
+
+def _objects_to_embed(objs, fh):
+    with h5py.File(fh, "w") as h5grp:
+        for i, obj in enumerate(objs):
+            # store string representation of the object
+            # that will serve as an identifier for the entire object.
+            # for sequences, this could be the sequence itself
+            # for molecules, this could be the SMILES string.
+            # The entries in this string representation can be used
+            # to index the row vectors in the embedding.
+            # For sequences, this is the positional index of the sequence.
+            # For molecules, this is the position index of atoms in the SMILES string.
+            arr = np.frombuffer(str(obj).encode("ascii"), dtype=np.uint8)
+            # Store the embedding itself. We are assuming that the
+            # embbedding is a 2D numpy array
+            emb = obj.embedding
+            # store the pointers that keep track of the start and
+            # end of the embedding for each object, as well as well as
+            # the corresponding string representation
+            if "idptr" in h5grp:
+                idptr_fh = h5grp["idptr"]
+                idptr_fh.resize((i + 1,))
```

Comment:
This likely will be expensive if done many times

Reply:
That has crossed my mind (I don't have intuition on large-scale benchmarks).
The only caveat here is that in order to allocate the entire array at once, we'd need to do a sweep over the entire generator to tally up the sum of the sequence lengths. It should be doable, I'd just need to figure out the schematics of iterating over a generator multiple times without triggering a `StopIteration`

Reply:
In `demux` if I remember, what we did was overallocate, trigger resize only when needed, and resize once more when done. This reduces the number of reallocs with a minor amount of book keeping. 

If these data are already in memory, then exhausting the generator won't impact memory, and would allow writing in bulk

Reply:
If the data are in memory, its not a problem. The issue is if we are doing with >100M proteins, that'll blow past 1TB of memory, so we would want to dynamically allocate these objects to disk in a streaming fashion.

Reply:
But this is on write, so at least the current item being yielded is in memory. Is it the case that we cannot assume other items in the generator are not in memory? Is the intended workflow something like the following?

```python
def transform(something):
    for stuff in skbio.read(something):
        cooler_stuff = do_the_thing(stuff)
        yield cooler_stuff

skbio.write(transform(a_thing), format='embedding')
```

?

Reply:
Basically yes, let me rewrite it into the typical workflow that I'm envisioning
```python
def encode_protein_from_llm(fasta_filepath):
    for protein in Protein.read(fasta_filepath):
        embedding = LLM(protein)   # outputs a ProteinEmbedding object
        yield embedding

embeddings = encode_protein_from_llm(fasta_file)
skbio.io.write(embeddings, format='embed', into=writable_emb_path)
```

Reply:
So it turns out that it'll be very complicated to clone generators (in order to figure out how much space to allocate).
https://stackoverflow.com/a/29835957

So let me proceed with the original suggestion and only resize only as necessary (this will amortize the runtime).

Reply:
I do recommend testing overallocate then reduce at the end as realloc is often expensive